W-21433121 agent network 2.0 content ie (#660)
mulesoft/docs-code-builder: agent-network/2.0/modules/ROOT/pages/af-example-it-investigation-broker.adoc

> [[example-building-an-it-investigation-broker]]

Reviewer: Does this work, putting an anchor on an H1? I just haven't seen it before.

Author: hmm, I'm gonna guess Claude did that

= Example: Building an IT Investigation Broker

> In this example, the goal is to use MuleSoft Vibes to build an IT Investigation agent that triages incoming support tickets, escalates critical issues, resolves common problems through cross-platform investigation, and handles licensing requests.

Reviewer: ```suggestion
Use MuleSoft Vibes to build an IT Investigation agent that triages incoming support tickets, escalates critical issues, resolves common problems through cross-platform investigation, and handles licensing requests.
```

In each phase, you interact with MuleSoft Vibes as it builds your agent network project. Here's how it works.

== Phase 1: Input Functional Requirements

> In this first phase, you define the functionality of the IT Investigation agent, which is to triage, investigate, and resolve support tickets. The agent must read incoming ticket descriptions (received via an A2A message with a Jira ticket ID), classify severity (High, Low, or too vague to classify) using an LLM's judgment, and then take the correct action. If the ticket is too vague, the agent first asks the user for clarification. 

Reviewer: ```suggestion
In this first phase, you define the functionality of the IT Investigation agent, which is to triage, investigate, and resolve support tickets. The agent reads incoming ticket descriptions (received via an A2A message with a Jira ticket ID), classifies severity (High, Low, or too vague to classify) using an LLM's judgment, and then takes the correct action. If the ticket is too vague, the agent first asks the user for clarification. 
```

> High-severity tickets must be immediately routed for escalation using an Escalation MCP tool. Low-severity tickets require multi-agent investigation, using a Help Center agent and a License Procurement agent. 

Reviewer: ```suggestion
High-severity tickets are immediately routed for escalation using an Escalation MCP tool. Low-severity tickets require multi-agent investigation, using a Help Center agent and a License Procurement agent. 
```

> This investigation path results in one of three deterministic outcomes: "Help Given," "License Given," or "Unresolved," the latter of which requires escalation to a human agent. Crucially, the severity classification and cross-platform investigation are open-ended, but the routing based on that classification and the final resolution must be deterministic.

Reviewer: I think Salesforce Agentforce style says to avoid calling humans agents. Can it just be human, or human rep?

You also select two MCP tools: the Escalation MCP Server for handling high-severity cases and the Jira MCP Server for updating ticket status. 

For the LLMs, you select OpenAI GPT-5.4 for reasoning and orchestration tasks, such as cross-platform triage. You register Gemini 2.5 Flash for fast, one-shot summaries and response message generation.

== Phase 3: Define the Broker (Agent Script)

In this phase, MuleSoft Vibes establishes the logical workflow for the IT Investigation agent by defining a series of interconnected nodes. The graph begins with an A2A trigger node (`ticketTrigger`) that receives the incoming ticket and Jira ID. This immediately flows into a generator node (`classifySeverity`), which uses LLM reasoning to classify the ticket as High or Low, and produces a structured output.

The deterministic `severityRouter` node then directs the process into two paths based on the classification output:

* High: An executor node calls the Escalation MCP tool. +

> * Low: An orchestrator node coordinates agents and tools to investigate, producing a structured outcome (e.g., `help_given`). A second router then directs the flow based on this outcome, with all paths concluding via an Echo node.

Reviewer: ```suggestion
* Low: An orchestrator node coordinates agents and tools to investigate, producing a structured outcome (for example, `help_given`). A second router then directs the flow based on this outcome, with all paths concluding via an Echo node.
```

|`helpSummary`
|generator
|--
|Gemini 2.5 Flash (default)
|--
|One-shot summary generation.

|`licenseSummary`
|generator
|--
|Gemini 2.5 Flash (default)
|--
|One-shot summary generation.

|`escalateTicket`
|executor
|Escalation MCP
|--
|1
|Deterministic action call (no LLM). Constraint: Can only escalate, not investigate or resolve.

|`escalateUnresolved`
|executor
|Escalation MCP
|--
|1
|Deterministic escalation for unresolved tickets (no LLM). Constraint: Can only escalate unresolved tickets after investigation has been exhausted.

|`All echo nodes`
|echo
|--
|--
|--
|Deterministic response formatting (no LLM).

|===

== Phase 5: Instruction Refinement

In this phase, MuleSoft Vibes writes and validates (with you) the precise natural language instructions for the LLM-powered nodes. The instructions for the `classifySeverity` node clearly define what constitutes High and Low severity, such as classifying a system outage as High severity. 

The `crossPlatformTriage` node instructions outline the required investigation steps: searching the Help Center, checking the License Procurement agent, and updating the Jira ticket. 

> Furthermore, these instructions define the two structured resolution outcomes the node must produce: `help_given` or `license_given`. A rigorous contradiction test is executed to ensure that no node's instructions accidentally override the deterministic routing logic defined in the graph. All instructions are verified to align with the overall graph structure and pass the cross-node checks.

Reviewer: ```suggestion
Furthermore, these instructions define the two structured resolution outcomes the node produces: `help_given` or `license_given`. MuleSoft Vibes runs a rigorous contradiction test to make sure that no node's instructions accidentally override the deterministic routing logic defined in the graph. MuleSoft Vibes verifies all instructions against the overall graph structure and cross-node checks.
```

== Phase 6: Final Topology Review

In this final phase, MuleSoft Vibes provides a comprehensive validation of the fully defined and instructed agent topology before deployment. The review confirms the crucial separation of concerns, ensuring LLM reasoning is used only for classification, with all subsequent routing being deterministic. 

The complete separation of the three severity paths - escalation, investigation, and asking for more info - is fully validated. The topology confirms that all possible paths terminate correctly at an echo response node to communicate the outcome to the original caller. Key operational parameters are also finalized, including maximum loop limits (5 for orchestration, 3 for classification) to prevent infinite loops. An edge case is defined, specifying that if the orchestration node cannot determine a resolution, the outcome defaults to `help_given`.

== IT Agent Broker Diagram

> The following diagram shows the workflow.

Reviewer: ```suggestion
This diagram shows the workflow.
```

image::af-it-broker-escalation-flow.png[Workflow diagram of the IT agent broker showing severity classification branching into escalation, investigation, and ask-for-more-info paths, each terminating in an echo response node.]

== IT Agent Project Reference

> The following shows the project files for the IT agent.

Reviewer: ```suggestion
These are the project files for the IT agent.
```

echo licenseResponse:
  kind: "a2a:response"
  task: a2a.task({
    state: "completed",
    message: a2a.message({
      messageId: uuid(),
      parts: [
        a2a.textPart(@generator.licenseSummary.output)
      ]
    }),
    metadata: None
  })

# -- UNRESOLVED PATH -----------------------------------------------------------
executor escalateUnresolved:
  description: "Escalates an unresolved low-severity ticket to a human agent."
  do: ->
    run @actions.escalate_ticket
      with ticket_id = @generator.classifySeverity.output.ticket_id
      with severity = "low"
      with reason = @orchestrator.crossPlatformTriage.output.summary
      with description = @request.payload.message.parts[0].text
  on_exit: ->
    transition to @echo.unresolvedResponse

echo unresolvedResponse:
  kind: "a2a:response"
  task: a2a.task({
    state: "completed",
    message: a2a.message({
      messageId: uuid(),
      parts: [
        a2a.textPart("Ticket " + @generator.classifySeverity.output.ticket_id + " could not be resolved automatically and has been escalated to a human agent. Summary: " + @orchestrator.crossPlatformTriage.output.summary)
      ]
    }),
    metadata: None
  })
----

> 

Reviewer: Add a See Also?
